File: csharp/examples/MakeGraphqlRequest.cs
// Example: Make an arbitrary GraphQL request to the Anvil API
//
// Feel free to copy and paste queries and mutations from the GraphQL reference
// docs into the functions in this script.
//
// * GraphQL guide: https://www.useanvil.com/docs/api/graphql
// * GraphQL ref docs: https://www.useanvil.com/docs/api/graphql/reference
// * Anvil C#/.NET client: https://www.nuget.org/packages/Anvil/
//
// This example is runnable as is, all you need to do is supply your own API key
// in the ANVIL_API_KEY environment variable.
//
// ANVIL_API_KEY=<yourAPIKey> dotnet run make-graphql-request

using Anvil.Client;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace AnvilExamples;

class MakeGraphqlRequest : RunnableBaseExample
{
    public static async Task<JObject> CallCurrentUserQuery(GraphQLClient client)
    {
        var query = @"query CurrentUser {
            currentUser {
                eid
                name
                organizations {
                    eid
                    slug
                    name
                    casts {
                        eid
                        name
                    }
                    welds {
                        eid
                        name
                    }
                }
            }
        }";
        return await client.SendQuery(query, null);
    }

    static async Task<JObject> CallWeldQuery(GraphQLClient client, string weldEid)
    {
        var query = @"
            query WeldQuery (
                $eid: String,
            ) {
                weld (
                    eid: $eid,
                ) {
                    eid
                    name
                    forges {
                        eid
                        slug
                        name
                    }
                }
            }";
        return await client.SendQuery(query, new {eid = weldEid});
    }

    /**
     * Main method that gets run from the `Program.cs` entrypoint.
     *
     * `apiKey` is provided through the `ANVIL_API_KEY` environment variable.
     */
    public override async Task Run(string apiKey)
    {
        Console.WriteLine("Make request");
        var client = new GraphQLClient(apiKey);
        var userResponse = await CallCurrentUserQuery(client);
        var userJson = JsonConvert.SerializeObject(userResponse);
        Console.WriteLine("Current user\n" + userJson);

        var firstWeld = userResponse["currentUser"]["organizations"][0]["welds"][0];
        var weldResponse = await CallWeldQuery(client, (string) firstWeld["eid"]);
        Console.WriteLine("First weld details:\n" + weldResponse);
    }

    public MakeGraphqlRequest(string apiKey) : base(apiKey)
    {
    }
}

File: ruby/examples/make_graphql_request.rb
require 'dotenv/load'
require "graphql/client"
require "graphql/client/http"

# Anvil GraphQL reference
# https://www.useanvil.com/docs/api/graphql/reference/#group-Operations-Queries

# More on GraphQL client usage at
# https://github.com/github/graphql-client

module AnvilAPI
  URL = 'https://app.useanvil.com/graphql'

  HTTP = GraphQL::Client::HTTP.new(URL) do
    def headers(context)
      apiKey = Base64.encode64("#{ENV['ANVIL_API_KEY']}:")
      {
        "Authorization": "Basic #{apiKey}".strip()
      }
    end
  end

  # Fetch latest schema on init, this will make a network request
  Schema = GraphQL::Client.load_schema(HTTP)

  # However, it's smart to dump this to a JSON file and load from disk
  #
  # Run it from a script or rake task
  #   GraphQL::Client.dump_schema(AnvilAPI::HTTP, "path/to/schema.json")
  #
  # Schema = GraphQL::Client.load_schema("path/to/schema.json")

  Client = GraphQL::Client.new(schema: Schema, execute: HTTP)
end

Queries = AnvilAPI::Client.parse <<-'GRAPHQL'
  query CurrentUserQuery {
    currentUser {
      eid
      name
      organizations {
        eid
        slug
        name
        casts {
          eid
        }
      }
    }
  }

  query CastQuery ($eid: String!) {
    cast (eid: $eid) {
      eid
      name
      isTemplate
    }
  }
GRAPHQL

# Run a query without any variables
result = AnvilAPI::Client.query(Queries::CurrentUserQuery)
result.data.to_h # Everything to a hash

puts("Current user:")
puts(result.data.current_user.eid)
puts(result.data.current_user.name)

# Run a query with variables
cast = result.data.current_user.organizations[0].casts[0]
result = AnvilAPI::Client.query(Queries::CastQuery, variables: { eid: cast.eid })

puts("\nFirst PDF Template (cast) object:")
puts(result.data.cast.eid)
puts(result.data.cast.name)
